content/advent-2015/etcd-distributed-key-value-store-with-grpc-http2.md

```diff
@@ -0,0 +1,362 @@
+		conn, err := grpc.Dial(endpoint, grpc.WithInsecure())
+		if err != nil {
+			panic(err)
+		}
+		conns[i] = conn
+	}
+	clients := make([]pb.KVClient, numClients)
+	for i := range clients {
+		clients[i] = pb.NewKVClient(conns[i%int(numConns)])
+	}
+
+	requests := make(chan *pb.PutRequest, len(keys))
+	done, errChan := make(chan struct{}), make(chan error)
+
+	for i := range clients {
+		go func(i int, requests chan *pb.PutRequest) {
+			for r := range requests {
+				if _, err := clients[i].Put(context.Background(), r); err != nil {
+					errChan <- err
+					return
+				}
+			}
+			done <- struct{}{}
+		}(i, requests)
+	}
+
+	st := time.Now()
+
+	for i := range keys {
+		r := &pb.PutRequest{
+			Key:   keys[i],
+			Value: vals[i],
+		}
+		requests <- r
+	}
+
+	close(requests)
+
+	cn := 0
+	for cn != len(clients) {
+		select {
+		case err := <-errChan:
+			panic(err)
+		case <-done:
+			cn++
+		}
+	}
+
+	tt := time.Since(st)
+	size := len(keys)
+	pt := tt / time.Duration(size)
+	log.Printf("GRPC took %v for %d requests with %d client(s) (%v per each).\n", tt, size, numClients, pt)
+}
+
+```
+
+## Benchmark results
+
+Tests send 300,000 requests to key/value stores. One uses `jsonrpc`, and the other uses `gRPC`. Both `jsonrpc` and `gRPC` code use only one TCP connection. And another interesting `gRPC` case with a single TCP connection but with multiple clients:
+
+|RPC|# of requests|# of clients|total time|per-request time|
+|---|---|---|---|:-:|
+|`jsonrpc`|300,000|1|8m7.270s|1.624ms|
+|gRPC|300,000|1|36.715s|122.383µs|
+|gRPC|300,000|100|7.167s|23.892µs|
+
+<br>
+And if compared on memory usage:
+
+|RPC|`jsonrpc`|gRPC|delta|
+|---|---------|----|:---:|
+|NsPerOp|487271046903|36716116701|-92.46%|
+|AllocsPerOp|32747687|25221256|-22.98%|
+|AllocedBytesPerOp|3182814152|1795122672|-43.60%|
+
+|RPC|`jsonrpc`|gRPC with 100 clients|delta|
+|---|---------|----|:---:|
+|NsPerOp|487271046903|7168591678|-98.53%|
+|AllocsPerOp|32747687|25230286|-22.96%|
+|AllocedBytesPerOp|3182814152|1795831944|-43.58%|
+
+|RPC|gRPC|gRPC with 100 clients|delta|
+|---|---------|----|:---:|
+|NsPerOp|36716116701|7168591678|-80.48%|
+|AllocsPerOp|25221256|25230286|+0.04%|
+|AllocedBytesPerOp|1795122672|1795831944|+0.04%|
+
+<br>
+As you see, gRPC is much faster and lighter than `jsonrpc`. Not only performant, but also gRPC is easier to reason about concurrency. HTTP/1.x requires multiple TCP connections for concurrent requests, while HTTP/2 can have multiple requests in-flight at the same time over one single TCP connections and still process them asynchronously. Tests above show that gRPC with multiple clients solely speeds up gRPC by 80%, without having to open multiple TCP connections.
+
+## How does etcd use gRPC?
+
+etcd already uses protocol buffers for all structured data, supporting gRPC as an experimental feature as of today. etcd team is actively working on gRPC integration to replace all etcd RPCs. Here's how etcd is structured to use gRPC:
```

Comment:
Here's how different parts of the etcd codebase leverage gRPC and Protocol Buffers: